analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs

```diff
@@ -582,6 +663,256 @@ void AsOperatorWithNonNullLiteral(int? i)
+
+        static void ModifyOutParam(out int? i) => i = null;
+        static void ModifyOutParamAndRead(out int? i1, int? i2) => i1 = i2;
+        static void ReadAndModifyOutParam(int? i1, out int? i2) => i2 = i1;
+    }
+
+    void RefParams(int? iParam)
+    {
+        iParam = null;
+        ModifyRefParam(ref iParam);
+        _ = iParam.Value;                                // Compliant, unknown after method call
+
+        int? iLocal = null;
+        ModifyRefParam(ref iLocal);
+        _ = iLocal.Value;                                // Compliant
+
+        iLocal = null;
+        ModifyRefParamAndRead(ref iLocal, iLocal.Value); // FN
+
+        iLocal = null;
+        ReadAndModifyRefParam(iLocal.Value, ref iLocal); // Noncompliant
+
+        nullableField = null;
+        ModifyRefParam(ref nullableField);
+        _ = nullableField.Value;                         // Compliant
+
+        static void ModifyRefParam(ref int? i) => i = null;
+        static void ModifyRefParamAndRead(ref int? i1, int? i2) => i1 = i2;
+        static void ReadAndModifyRefParam(int? i1, ref int? i2) => i2 = i1;
+    }
+}
+
+class MutableField
+{
+    int? theField;
+
+    void Basics()
+    {
+        _ = theField.Value;          // Compliant, unknown
+
+        theField = null;
+        _ = theField.Value;          // Noncompliant, empty
+        _ = theField.Value;          // Compliant, when reached the ".Value" above implies theField is not null
+    }
+
+    void LocalFunctions()
+    {
+        theField = 42;
+        LocalFunctionChangingTheField();
+        _ = theField.Value;          // Compliant, unknown
+
+        theField = null;
+        LocalFunctionChangingTheField();
+        _ = theField.Value;          // Noncompliant, FP, local functions don't reset the state
+
+        theField = 42;
+        LocalFunctionNotChangingTheField();
+        _ = theField.Value;          // Compliant, unknown
+
+        void LocalFunctionChangingTheField() => theField = null;
+        void LocalFunctionNotChangingTheField() { }
+    }
+
+    void Methods()
+    {
+        theField = 42;
+        MethodChangingTheField();
+        _ = theField.Value;          // Compliant, unknown
+
+        theField = null;
+        MethodChangingTheField();
+        _ = theField.Value;          // Compliant, instance methods reset the state
+
+        theField = 42;
+        MethodNotChangingTheField();
+        _ = theField.Value;          // Compliant, unknown
+
+        void LocalFunctionChangingTheField() => theField = null;
+        void LocalFunctionNotChangingTheField() { }
+    }
+
+    void MethodChangingTheField() => theField = null;
+
+    void MethodNotChangingTheField() { }
+
+    IEnumerable<int> WithYield()
+    {
+        _ = theField.Value;          // Compliant, unknown
+        yield return 42;
+
+        theField = null;
+        yield return 42;
+        _ = theField.Value;          // Noncompliant, FP, should be unknown
```

Comment:
@pavel-mikula-sonarsource another one for your attention. Where do we track these? `yield return` should behave exactly like `await` as both turn the method into a state machine with a re-entry at a later point in execution.

Reply:
We don't (yet). 
I think there's a difference between await and yield return. Await can mess up your local variables and parameters, while yield return can mess up only external state (fields), like a local method invocation.

So I'd propose to threat it like a local method invocation to reset only some state.